File: python/journal/reader_access.py
"""Bounded reader byte access for journal files."""

from __future__ import annotations

import collections
import contextlib
import mmap
import os

from ._platform_io import read_at, read_at_uses_pread


READER_ACCESS_AUTO = "auto"
READER_ACCESS_MMAP = "mmap"
READER_ACCESS_READ_AT = "read-at"
READER_BOUNDS_LIVE = "live"
READER_BOUNDS_SNAPSHOT = "snapshot"

DEFAULT_WINDOW_SIZE = 32 * 1024 * 1024
DEFAULT_MAX_WINDOWS = 4
DEFAULT_MAX_ROW_ARENA_BYTES = 256 * 1024 * 1024
DEFAULT_ROW_ARENA_SEGMENT_BYTES = 1024 * 1024


class ReaderOptions:
    def __init__(
        self,
        access_mode=READER_ACCESS_AUTO,
        window_size=DEFAULT_WINDOW_SIZE,
        max_windows=DEFAULT_MAX_WINDOWS,
        max_row_arena_bytes=DEFAULT_MAX_ROW_ARENA_BYTES,
        max_retired_windows=None,
        row_arena_segment_bytes=DEFAULT_ROW_ARENA_SEGMENT_BYTES,
        bounds=READER_BOUNDS_LIVE,
    ):
        self.access_mode = _normalize_access_mode(access_mode)
        self.window_size = int(window_size)
        self.max_windows = int(max_windows)
        self.max_row_arena_bytes = int(max_row_arena_bytes)
        self.max_retired_windows = (
            self.max_windows if max_retired_windows is None else int(max_retired_windows)
        )
        self.row_arena_segment_bytes = int(row_arena_segment_bytes)
        self.bounds = _normalize_bounds(bounds)
        self._validate()

    def _validate(self):
        if self.window_size <= 0:
            raise ValueError("window_size must be positive")
        if self.max_windows <= 0:
            raise ValueError("max_windows must be positive")
        if self.max_row_arena_bytes < 0:
            raise ValueError("max_row_arena_bytes must not be negative")
        if self.max_retired_windows < 0:
            raise ValueError("max_retired_windows must not be negative")
        if self.row_arena_segment_bytes <= 0:
            raise ValueError("row_arena_segment_bytes must be positive")


def default_reader_options():
    return ReaderOptions()


def _normalize_access_mode(access_mode):
    text = str(access_mode).lower().replace("_", "-")
    aliases = {
        "auto": READER_ACCESS_AUTO,
        "mmap": READER_ACCESS_MMAP,
        "readat": READER_ACCESS_READ_AT,
        "read-at": READER_ACCESS_READ_AT,
        "pread": READER_ACCESS_READ_AT,
    }
    try:
        return aliases[text]
    except KeyError as err:
        raise ValueError(f"unsupported reader access mode: {access_mode}") from err


def _normalize_bounds(bounds):
    text = str(bounds).lower().replace("_", "-")
    aliases = {
        "live": READER_BOUNDS_LIVE,
        "snapshot": READER_BOUNDS_SNAPSHOT,
    }
    try:
        return aliases[text]
    except KeyError as err:
        raise ValueError(f"unsupported reader bounds mode: {bounds}") from err


def open_reader_accessor(path, options=None):
    opts = options if isinstance(options, ReaderOptions) else ReaderOptions() if options is None else options
    if not isinstance(opts, ReaderOptions):
        raise TypeError("options must be ReaderOptions or None")

    fd = os.open(path, os.O_RDONLY | getattr(os, "O_BINARY", 0))
    try:
        if opts.access_mode == READER_ACCESS_READ_AT:
            return _ReadAtAccessor(fd, opts, selected=READER_ACCESS_READ_AT)
        if opts.access_mode == READER_ACCESS_MMAP:
            return _MmapAccessor(fd, opts, selected=READER_ACCESS_MMAP)
        try:
            return _MmapAccessor(fd, opts, selected=READER_ACCESS_MMAP)
        except Exception as mmap_err:
            accessor = _ReadAtAccessor(
                fd,
                opts,
                selected=READER_ACCESS_READ_AT,
                fallback_reason=str(mmap_err),
            )
            return accessor
    except Exception:
        with contextlib.suppress(Exception):
            os.close(fd)
        raise


class _StatsMixin:
    def selected_access_mode(self):
        return self._selected

    def stats(self):
        return {
            "selected_backend": self._selected,
            "fallback_reason": self._fallback_reason,
            "visible_size": self._visible_size,
            "window_size": self._options.window_size,
            "max_windows": self._options.max_windows,
            "current_windows": len(self._windows),
            "mapped_bytes": self._mapped_bytes(),
            "read_buffer_bytes": self._read_buffer_bytes(),
            "row_pinned_windows": sum(1 for window in self._windows.values() if window.row_pinned),
            "retired_windows": len(getattr(self, "_retired", [])),
            "retired_bytes": sum(window.length for window in getattr(self, "_retired", [])),
            "row_arena_current_bytes": self._row_arena.current_bytes,
            "row_arena_peak_bytes": self._row_arena.peak_bytes,
            "row_arena_limit_bytes": self._row_arena.limit_bytes,
            "row_arena_segment_bytes": self._row_arena.segment_bytes,
            "row_arena_active_segments": len(self._row_arena.segments),
            "read_at_uses_pread": read_at_uses_pread(),
            "temp_copy_count": self._temp_copy_count,
            "window_miss_count": self._window_miss_count,
            "eviction_count": self._eviction_count,
        }


class _BaseAccessor(_StatsMixin):
    def __init__(self, fd, options, selected, fallback_reason=None):
        self._fd = fd
        self._options = options
        self._selected = selected
        self._fallback_reason = fallback_reason
        self._visible_size = os.fstat(fd).st_size
        self._windows = collections.OrderedDict()
        self._scratch = None
        self._closed = False
        self._temp_copy_count = 0
        self._window_miss_count = 0
        self._eviction_count = 0
        self._row_arena = _RowArena(options.max_row_arena_bytes, options.row_arena_segment_bytes)

    def size(self):
        return self._visible_size

    def fd(self):
        return self._fd

    def bounds_mode(self):
        return self._options.bounds

    def read_bytes(self, offset, size):
        return bytes(self.temp_view(offset, size))

    def row_bytes(self, data):
        return self._row_arena.append(data)

    def u8(self, offset):
        return self.temp_view(offset, 1)[0]

    def u32(self, offset):
        return int.from_bytes(self.temp_view(offset, 4), "little")

    def u64(self, offset):
        return int.from_bytes(self.temp_view(offset, 8), "little")

    def clear_row(self):
        for window in list(self._windows.values()):
            window.row_pinned = False
        self._row_arena.clear()
        self._evict_to_budget()
        self._retry_retired()

    def snapshot_visible_bounds(self):
        return self._visible_size

    def restore_visible_bounds(self, snapshot):
        self._visible_size = int(snapshot)

    def refresh_visible_bounds(self):
        if self._options.bounds == READER_BOUNDS_SNAPSHOT:
            return False, self._visible_size
        size = os.fstat(self._fd).st_size
        changed = size != self._visible_size
        self._visible_size = size
        return changed, size

    def close(self):
        if self._closed:
            return
        self.clear_row()
        for base in list(self._windows):
            self._close_window(base)
        self._retry_retired()
        with contextlib.suppress(Exception):
            os.close(self._fd)
        self._closed = True

    def _check_range(self, offset, size):
        offset = int(offset)
        size = int(size)
        if offset < 0 or size < 0:
            raise ValueError("negative offset or size")
        if offset + size > self._visible_size:
            raise ValueError("read exceeds visible file bounds")
        return offset, size

    def _scratch_view(self, offset, size):
        offset, size = self._check_range(offset, size)
        data = read_at(self._fd, size, offset)
        if len(data) != size:
            raise ValueError("short read")
        self._scratch = data
        self._temp_copy_count += 1
        return memoryview(data)

    def _mapped_bytes(self):
        return 0

    def _read_buffer_bytes(self):
        return 0

    def _retry_retired(self):
        return None


class _MmapWindow:
    def __init__(self, base, length, mapping):
        self.base = base
        self.length = length
        self.mapping = mapping
        self.view = memoryview(mapping)
        self.row_pinned = False


class _MmapAccessor(_BaseAccessor):
    def __init__(self, fd, options, selected, fallback_reason=None):
        super().__init__(fd, options, selected, fallback_reason)
        self._granularity = int(getattr(mmap, "ALLOCATIONGRANULARITY", mmap.PAGESIZE))
        self._retired = []
        if self._visible_size > 0:
            self.temp_view(0, min(self._visible_size, min(options.window_size, 1)))

    def temp_view(self, offset, size):
        return self._view(offset, size, row=False)

    def row_view(self, offset, size):
        try:
            return self._view(offset, size, row=True)
        except RuntimeError:
            return self._row_arena.append(bytes(self.temp_view(offset, size)))

    def _view(self, offset, size, row):
        offset, size = self._check_range(offset, size)
        if size == 0:
            return memoryview(b"")
        if size > self._options.window_size:
            if row:
                return self._row_arena.append(bytes(self._scratch_view(offset, size)))
            return self._scratch_view(offset, size)

        base = (offset // self._granularity) * self._granularity
        window = self._windows.get(base)
        if window is None or offset + size > window.base + window.length:
            if window is not None and window.row_pinned:
                if row:
                    return self._row_arena.append(bytes(self._scratch_view(offset, size)))
                return self._scratch_view(offset, size)
            if len(self._windows) >= self._options.max_windows and self._all_windows_row_pinned():
                if row:
                    return self._row_arena.append(bytes(self._scratch_view(offset, size)))
                return self._scratch_view(offset, size)
            window = self._map_window(base, offset, size)
        else:
            self._windows.move_to_end(base)
        if row:
            window.row_pinned = True
        start = offset - window.base
        return window.view[start:start + size]

    def _map_window(self, base, offset, size):
        existing = self._windows.get(base)
        if existing is not None:
            if existing.row_pinned:
                raise RuntimeError("cannot replace row-pinned mmap window")
            self._close_window(base)
            self._eviction_count += 1
        else:
            self._evict_to_budget(extra_needed=1)
        length = min(self._options.window_size + (offset - base), self._visible_size - base)
        if offset + size > base + length:
            length = offset + size - base
        mapping = mmap.mmap(self._fd, length, access=mmap.ACCESS_READ, offset=base)
        window = _MmapWindow(base, length, mapping)
        self._windows[base] = window
        self._window_miss_count += 1
        self._evict_to_budget()
        return window

    def _all_windows_row_pinned(self):
        return bool(self._windows) and all(window.row_pinned for window in self._windows.values())

    def _evict_to_budget(self, extra_needed=0):
        while len(self._windows) + extra_needed > self._options.max_windows:
            for base, window in list(self._windows.items()):
                if not window.row_pinned:
                    self._close_window(base)
                    self._eviction_count += 1
                    break
            else:
                return

    def _close_window(self, base):
        window = self._windows.pop(base, None)
        if window is None:
            return
        with contextlib.suppress(Exception):
            window.view.release()
        try:
            window.mapping.close()
        except BufferError:
            self._retire_window(window)

    def _retire_window(self, window):
        self._retired.append(window)
        if len(self._retired) > self._options.max_retired_windows:
            raise RuntimeError("too many retired mmap windows; row views were retained past row lifetime")

    def _retry_retired(self):
        remaining = []
        for window in self._retired:
            with contextlib.suppress(Exception):
                window.view.release()
            try:
                window.mapping.close()
            except BufferError:
                remaining.append(window)
        self._retired = remaining

    def _mapped_bytes(self):
        return sum(window.length for window in self._windows.values())


class _ReadAtWindow:
    def __init__(self, base, data):
        self.base = base
        self.data = data
        self.length = len(data)
        self.view = memoryview(data)
        self.row_pinned = False


class _ReadAtAccessor(_BaseAccessor):
    def temp_view(self, offset, size):
        return self._view(offset, size, row=False)

    def row_view(self, offset, size):
        try:
            return self._view(offset, size, row=True)
        except RuntimeError:
            return self._row_arena.append(bytes(self.temp_view(offset, size)))

    def _view(self, offset, size, row):
        offset, size = self._check_range(offset, size)
        if size == 0:
            return memoryview(b"")
        if size > self._options.window_size:
            if row:
                return self._row_arena.append(bytes(self._scratch_view(offset, size)))
            return self._scratch_view(offset, size)
        base = (offset // self._options.window_size) * self._options.window_size
        window = self._windows.get(base)
        if window is None or offset + size > window.base + window.length:
            if window is not None and window.row_pinned:
                if row:
                    return self._row_arena.append(bytes(self._scratch_view(offset, size)))
                return self._scratch_view(offset, size)
            if len(self._windows) >= self._options.max_windows and self._all_windows_row_pinned():
                if row:
                    return self._row_arena.append(bytes(self._scratch_view(offset, size)))
                return self._scratch_view(offset, size)
            window = self._read_window(base, offset, size)
        else:
            self._windows.move_to_end(base)
        if row:
            window.row_pinned = True
        start = offset - window.base
        return window.view[start:start + size]

    def _read_window(self, base, offset, size):
        existing = self._windows.get(base)
        if existing is not None:
            if existing.row_pinned:
                raise RuntimeError("cannot replace row-pinned read-at window")
            self._close_window(base)
            self._eviction_count += 1
        else:
            self._evict_to_budget(extra_needed=1)
        length = min(self._options.window_size, self._visible_size - base)
        if offset + size > base + length:
            length = offset + size - base
        data = read_at(self._fd, length, base)
        if len(data) != length:
            raise ValueError("short read")
        window = _ReadAtWindow(base, data)
        self._windows[base] = window
        self._window_miss_count += 1
        self._evict_to_budget()
        return window

    def _all_windows_row_pinned(self):
        return bool(self._windows) and all(window.row_pinned for window in self._windows.values())

    def _evict_to_budget(self, extra_needed=0):
        while len(self._windows) + extra_needed > self._options.max_windows:
            for base, window in list(self._windows.items()):
                if not window.row_pinned:
                    self._windows.pop(base, None)
                    self._eviction_count += 1
                    break
            else:
                return

    def _close_window(self, base):
        self._windows.pop(base, None)

    def _read_buffer_bytes(self):
        return sum(window.length for window in self._windows.values())


class _RowArena:
    def __init__(self, limit_bytes, segment_bytes):
        self.limit_bytes = int(limit_bytes)
        self.segment_bytes = int(segment_bytes)
        self.segments = []
        self.current_bytes = 0
        self.peak_bytes = 0
        self._used_in_current = 0

    def append(self, data):
        data = bytes(data)
        size = len(data)
        if self.current_bytes + size > self.limit_bytes:
            raise RuntimeError("row arena limit exceeded")
        if not self.segments or self._remaining_current() < size:
            segment_size = max(self.segment_bytes, size)
            self.segments.append(bytearray(segment_size))
            self._used_in_current = 0
        segment = self.segments[-1]
        start = self._used_in_current
        segment[start:start + size] = data
        self._used_in_current += size
        self.current_bytes += size
        self.peak_bytes = max(self.peak_bytes, self.current_bytes)
        return memoryview(segment)[start:start + size]

    def clear(self):
        self.segments = []
        self.current_bytes = 0
        self._used_in_current = 0

    def _remaining_current(self):
        if not self.segments:
            return 0
        return len(self.segments[-1]) - self._used_in_current

File: experiments/node/src/lib/reader-access.js
import { closeSync, fstatSync, openSync, readSync } from 'node:fs';

export const READER_ACCESS_AUTO = 'auto';
export const READER_ACCESS_READ_AT = 'read-at';
export const READER_ACCESS_MMAP = 'mmap';
export const READER_BOUNDS_LIVE = 'live';
export const READER_BOUNDS_SNAPSHOT = 'snapshot';

export const DEFAULT_WINDOW_SIZE_BYTES = 32 * 1024 * 1024;
export const DEFAULT_MAX_WINDOWS = 4;
export const DEFAULT_MAX_ROW_ARENA_BYTES = 256 * 1024 * 1024;
export const DEFAULT_ROW_ARENA_SEGMENT_BYTES = 1024 * 1024;

const EMPTY_BUFFER = Buffer.alloc(0);

export class UnsupportedAccessModeError extends Error {
  constructor(mode) {
    super(`unsupported reader access mode for Node.js core: ${mode}`);
    this.name = 'UnsupportedAccessModeError';
    this.accessMode = mode;
  }
}

export function normalizeReaderOptions(options = {}) {
  const source = options ?? {};
  const accessMode = normalizeAccessMode(source.accessMode ?? source.mmapStrategy ?? READER_ACCESS_AUTO);
  const bounds = normalizeBounds(source.bounds ?? READER_BOUNDS_LIVE);
  const windowSizeBytes = normalizePositiveInteger(
    source.windowSizeBytes ?? source.windowSize ?? DEFAULT_WINDOW_SIZE_BYTES,
    'windowSizeBytes',
  );
  const maxWindows = normalizePositiveInteger(source.maxWindows ?? DEFAULT_MAX_WINDOWS, 'maxWindows');
  const maxRowArenaBytes = normalizeNonNegativeInteger(
    source.maxRowArenaBytes ?? DEFAULT_MAX_ROW_ARENA_BYTES,
    'maxRowArenaBytes',
  );
  const rowArenaSegmentBytes = normalizePositiveInteger(
    source.rowArenaSegmentBytes ?? DEFAULT_ROW_ARENA_SEGMENT_BYTES,
    'rowArenaSegmentBytes',
  );
  const zstdTimeoutMs = normalizePositiveInteger(source.zstdTimeoutMs ?? 600000, 'zstdTimeoutMs');

  return {
    accessMode,
    bounds,
    windowSizeBytes,
    maxWindows,
    maxRowArenaBytes,
    rowArenaSegmentBytes,
    zstdTimeoutMs,
  };
}

export function withSnapshotBounds(options = {}) {
  return { ...normalizeReaderOptions(options), bounds: READER_BOUNDS_SNAPSHOT };
}

export function openReaderAccessor(path, options = {}) {
  const opts = normalizeReaderOptions(options);
  if (opts.accessMode === READER_ACCESS_MMAP) {
    throw new UnsupportedAccessModeError(opts.accessMode);
  }
  const fd = openSync(path, 'r');
  try {
    return new ReadAtAccessor(fd, opts);
  } catch (error) {
    try { closeSync(fd); } catch {
      // Preserve the original open failure.
    }
    throw error;
  }
}

function normalizeAccessMode(value) {
  const text = String(value ?? READER_ACCESS_AUTO).toLowerCase().replaceAll('_', '-');
  if (text === 'auto' || text === 'buffer') return READER_ACCESS_AUTO;
  if (text === 'readat' || text === 'read-at' || text === 'pread') return READER_ACCESS_READ_AT;
  if (text === 'mmap') return READER_ACCESS_MMAP;
  throw new Error(`unsupported reader access mode: ${value}`);
}

function normalizeBounds(value) {
  const text = String(value ?? READER_BOUNDS_LIVE).toLowerCase().replaceAll('_', '-');
  if (text === READER_BOUNDS_LIVE) return READER_BOUNDS_LIVE;
  if (text === READER_BOUNDS_SNAPSHOT) return READER_BOUNDS_SNAPSHOT;
  throw new Error(`unsupported reader bounds mode: ${value}`);
}

function normalizePositiveInteger(value, name) {
  const number = Number(value);
  if (!Number.isSafeInteger(number) || number <= 0) {
    throw new Error(`${name} must be a positive safe integer`);
  }
  return number;
}

function normalizeNonNegativeInteger(value, name) {
  const number = Number(value);
  if (!Number.isSafeInteger(number) || number < 0) {
    throw new Error(`${name} must be a non-negative safe integer`);
  }
  return number;
}

class ReadAtAccessor {
  constructor(fd, options) {
    this.fd = fd;
    this.options = options;
    this.visibleSize = fstatSync(fd).size;
    this.windows = new Map();
    this.rowArena = new RowArena(options.maxRowArenaBytes, options.rowArenaSegmentBytes);
    this.closed = false;
    this.scratch = Buffer.alloc(8);

    this.statsValue = {
      requestedAccessMode: options.accessMode,
      selectedAccessMode: READER_ACCESS_READ_AT,
      selectedBackend: READER_ACCESS_READ_AT,
      fallbackReason: options.accessMode === READER_ACCESS_AUTO ? 'Node.js core has no mmap API' : '',
      bounds: options.bounds,
      visibleSize: this.visibleSize,
      windowSizeBytes: options.windowSizeBytes,
      maxWindows: options.maxWindows,
      windowsCreated: 0,
      windowHits: 0,
      windowMisses: 0,
      evictions: 0,
      pinnedWindows: 0,
      readBufferBytes: 0,
      rowArenaPeakBytes: 0,
      tempCopyBytes: 0,
      tempCopyCount: 0,
      shortReads: 0,
      readSyncUsesPosition: true,
    };
  }

  size() {
    return this.visibleSize;
  }

  stats() {
    this.statsValue.visibleSize = this.visibleSize;
    this.statsValue.pinnedWindows = Array.from(this.windows.values()).filter((window) => window.rowPinned).length;
    this.statsValue.readBufferBytes = Array.from(this.windows.values())
      .reduce((total, window) => total + window.buffer.length, 0);
    this.statsValue.rowArenaPeakBytes = this.rowArena.peakBytes;
    return { ...this.statsValue };
  }

  snapshotVisibleBounds() {
    return this.visibleSize;
  }

  restoreVisibleBounds(size) {
    this.visibleSize = Number(size);
  }

  refreshVisibleBounds() {
    if (this.options.bounds === READER_BOUNDS_SNAPSHOT) return false;
    const nextSize = fstatSync(this.fd).size;
    const changed = nextSize !== this.visibleSize;
    this.visibleSize = nextSize;
    this._dropUnpinnedWindows();
    return changed;
  }

  u8(offset) {
    return this.tempView(offset, 1)[0];
  }

  u32(offset) {
    return this.tempView(offset, 4).readUInt32LE(0);
  }

  u64(offset) {
    return this.tempView(offset, 8).readBigUInt64LE(0);
  }

  readBytes(offset, size) {
    return Buffer.from(this.tempView(offset, size));
  }

  tempView(offset, size) {
    return this._view(offset, size, false);
  }

  rowView(offset, size) {
    return this._view(offset, size, true);
  }

  rowBytes(data) {
    return this.rowArena.append(data);
  }

  updateHmac(hmac, offset, size, chunkSize = 1 << 20) {
    let pos = Number(offset);
    let remaining = Number(size);
    if (remaining < 0) throw new Error('negative HMAC range size');
    while (remaining > 0) {
      const chunk = Math.min(remaining, chunkSize);
      hmac.update(this.tempView(pos, chunk));
      pos += chunk;
      remaining -= chunk;
    }
  }

  clearRow() {
    for (const window of this.windows.values()) window.rowPinned = false;
    this.rowArena.clear();
    this._evictToBudget();
  }

  close() {
    if (this.closed) return;
    this.clearRow();
    this.windows.clear();
    closeSync(this.fd);
    this.closed = true;
  }

  _view(offset, size, row) {
    const { pos, len } = this._checkRange(offset, size);
    if (len === 0) return EMPTY_BUFFER;
    if (len > this.options.windowSizeBytes) {
      const copy = this._scratchRead(pos, len, row);
      return row ? this.rowBytes(copy) : copy;
    }

    const base = Math.floor(pos / this.options.windowSizeBytes) * this.options.windowSizeBytes;
    if (pos + len > base + this.options.windowSizeBytes) {
      const copy = this._scratchRead(pos, len, row);
      return row ? this.rowBytes(copy) : copy;
    }
    const window = this._windowFor(base, pos, len, row);
    const start = pos - window.base;
    if (row) window.rowPinned = true;
    return window.buffer.subarray(start, start + len);
  }

  _checkRange(offset, size) {
    const pos = Number(offset);
    const len = Number(size);
    if (!Number.isSafeInteger(pos) || !Number.isSafeInteger(len) || pos < 0 || len < 0) {
      throw new Error('invalid read range');
    }
    if (pos + len > this.visibleSize) {
      throw new Error('read exceeds visible file bounds');
    }
    return { pos, len };
  }

  _windowFor(base, offset, size, row) {
    const existing = this.windows.get(base);
    if (existing && offset + size <= existing.base + existing.buffer.length) {
      this.windows.delete(base);
      this.windows.set(base, existing);
      this.statsValue.windowHits++;
      return existing;
    }
    if (existing?.rowPinned) {
      if (row) return { base: offset, buffer: this.rowBytes(this._scratchRead(offset, size, true)), rowPinned: true };
      return { base: offset, buffer: this._scratchRead(offset, size, false), rowPinned: false };
    }
    if (this.windows.size >= this.options.maxWindows && this._allWindowsPinned()) {
      if (row) return { base: offset, buffer: this.rowBytes(this._scratchRead(offset, size, true)), rowPinned: true };
      return { base: offset, buffer: this._scratchRead(offset, size, false), rowPinned: false };
    }
    if (existing) {
      this.windows.delete(base);
      this.statsValue.evictions++;
    } else {
      this._evictToBudget(1);
    }
    const length = Math.min(this.options.windowSizeBytes, this.visibleSize - base);
    const buffer = Buffer.alloc(length);
    const bytes = readSync(this.fd, buffer, 0, length, base);
    if (bytes !== length) {
      this.statsValue.shortReads++;
      throw new Error('short read before visible file size');
    }
    const window = { base, buffer, rowPinned: false };
    this.windows.set(base, window);
    this.statsValue.windowsCreated++;
    this.statsValue.windowMisses++;
    this._evictToBudget();
    return window;
  }

  _scratchRead(offset, size, stable) {
    const buffer = size <= this.scratch.length ? this.scratch.subarray(0, size) : Buffer.alloc(size);
    const bytes = readSync(this.fd, buffer, 0, size, offset);
    if (bytes !== size) {
      this.statsValue.shortReads++;
      throw new Error('short read before visible file size');
    }
    this.statsValue.tempCopyBytes += size;
    this.statsValue.tempCopyCount++;
    return stable && size <= this.scratch.length ? Buffer.from(buffer) : buffer;
  }

  _allWindowsPinned() {
    return this.windows.size > 0 && Array.from(this.windows.values()).every((window) => window.rowPinned);
  }

  _dropUnpinnedWindows() {
    for (const [base, window] of this.windows) {
      if (!window.rowPinned) {
        this.windows.delete(base);
        this.statsValue.evictions++;
      }
    }
  }

  _evictToBudget(extraNeeded = 0) {
    while (this.windows.size + extraNeeded > this.options.maxWindows) {
      let evicted = false;
      for (const [base, window] of this.windows) {
        if (!window.rowPinned) {
          this.windows.delete(base);
          this.statsValue.evictions++;
          evicted = true;
          break;
        }
      }
      if (!evicted) return;
    }
  }
}

class RowArena {
  constructor(limitBytes, segmentBytes) {
    this.limitBytes = limitBytes;
    this.segmentBytes = segmentBytes;
    this.segments = [];
    this.currentBytes = 0;
    this.peakBytes = 0;
    this.usedInCurrent = 0;
  }

  append(data) {
    const src = bufferView(data);
    if (this.currentBytes + src.length > this.limitBytes) {
      throw new Error('row arena limit exceeded');
    }
    if (this.segments.length === 0 || this._remaining() < src.length) {
      this.segments.push(Buffer.alloc(Math.max(this.segmentBytes, src.length)));
      this.usedInCurrent = 0;
    }
    const segment = this.segments[this.segments.length - 1];
    const start = this.usedInCurrent;
    src.copy(segment, start);
    this.usedInCurrent += src.length;
    this.currentBytes += src.length;
    this.peakBytes = Math.max(this.peakBytes, this.currentBytes);
    return segment.subarray(start, start + src.length);
  }

  clear() {
    this.segments = [];
    this.currentBytes = 0;
    this.usedInCurrent = 0;
  }

  _remaining() {
    if (this.segments.length === 0) return 0;
    return this.segments[this.segments.length - 1].length - this.usedInCurrent;
  }
}

function bufferView(data) {
  if (Buffer.isBuffer(data)) return data;
  if (data instanceof Uint8Array) {
    return Buffer.from(data.buffer, data.byteOffset, data.byteLength);
  }
  return Buffer.from(data);
}
